modules/navier_stokes/unit/src/TestReconstruction.C

```diff
@@ -0,0 +1,207 @@
+
+  MultiMooseEnum coord_type_enum("XYZ RZ RSPHERICAL", "XYZ");
+  coord_type_enum = (coord_type == Moose::COORD_XYZ) ? "XYZ" : "RZ";
+
+  std::vector<unsigned int> num_elem = {64, 128, 256};
+  std::vector<Real> weller_errors;
+  std::vector<Real> moukalled_errors;
+  std::vector<Real> tano_errors;
+  std::vector<Real> tano_errors_twice;
+  std::vector<Real> h(num_elem.size());
+  for (const auto i : index_range(num_elem))
+    h[i] = 1. / num_elem[i];
+
+  for (const auto i : index_range(num_elem))
+  {
+    const auto nx = num_elem[i];
+    auto app = AppFactory::createAppShared("NavierStokesApp", 1, (char **)argv);
+    auto * factory = &app->getFactory();
+    std::string mesh_type = "MeshGeneratorMesh";
+
+    std::shared_ptr<MeshGeneratorMesh> mesh;
+    {
+      InputParameters params = factory->getValidParams(mesh_type);
+      mesh = factory->create<MeshGeneratorMesh>(mesh_type, "moose_mesh", params);
+    }
+
+    app->actionWarehouse().mesh() = mesh;
+
+    {
+      std::unique_ptr<MeshBase> lm_mesh;
+      InputParameters params = factory->getValidParams("GeneratedMeshGenerator");
+      params.set<unsigned int>("nx") = nx;
+      params.set<unsigned int>("ny") = nx;
+      params.set<MooseEnum>("dim") = "2";
+      auto mesh_gen =
+          factory->create<GeneratedMeshGenerator>("GeneratedMeshGenerator", "mesh_gen", params);
+      lm_mesh = mesh_gen->generate();
+      mesh->setMeshBase(std::move(lm_mesh));
+    }
+
+    mesh->prepare();
+    mesh->setCoordSystem({}, coord_type_enum);
+    mooseAssert(mesh->getAxisymmetricRadialCoord() == 0,
+                "This should be 0 because we haven't set anything.");
+    const auto & all_fi = mesh->allFaceInfo();
+    mesh->computeFaceInfoFaceCoords();
+    std::vector<const FaceInfo *> faces(all_fi.size());
+    for (const auto i : index_range(all_fi))
+      faces[i] = &all_fi[i];
+
+    auto & lm_mesh = mesh->getMesh();
+
+    CellCenteredMapFunctor<RealVectorValue, std::unordered_map<dof_id_type, RealVectorValue>> u(
+        *mesh);
+    for (auto * const elem : lm_mesh.active_element_ptr_range())
+    {
+      const auto centroid = elem->vertex_average();
+      const auto value = RealVectorValue(-std::sin(centroid(0)) * std::cos(centroid(1)),
+                                         std::cos(centroid(0)) * std::sin(centroid(1)));
+      u[elem->id()] = value;
+    }
+
+    CellCenteredMapFunctor<RealVectorValue, std::unordered_map<dof_id_type, RealVectorValue>>
+        up_weller(*mesh);
+    std::unordered_map<dof_id_type, RealVectorValue> up_moukalled;
+    CellCenteredMapFunctor<RealVectorValue, std::unordered_map<dof_id_type, RealVectorValue>>
+        up_tano(*mesh);
+
+    for (const auto & fi : all_fi)
+    {
+      auto moukalled_reconstruct = [&fi](auto & functor, auto & container) {
+        auto face = Moose::FV::makeCDFace(fi);
+        const RealVectorValue uf(functor(face));
+        const Point surface_vector = fi.normal() * fi.faceArea();
```

Comment:
I don't think so. I don't find his reconstruction of B2 very compelling. For instance, equation 15.209 is wrong for cylindrical coordinates because gradient is not equivalent to divergence (with identity matrix) in that case. And then in 15.210 the simplification on the third line simply doesn't hold if you make Sf include 2*pi*r, e.g. the sum of the face areas * normals will not zero.

Reply:
To be frank, I'm going to find a new book after this PR